llvm/lib/Target/AArch64/AArch64SelectionDAGInfo.cpp

```diff
@@ -76,12 +76,66 @@ SDValue AArch64SelectionDAGInfo::EmitMOPS(AArch64ISD::NodeType SDOpcode,
   }
 }
 
+SDValue AArch64SelectionDAGInfo::EmitSpecializedLibcall(
+    SelectionDAG &DAG, const SDLoc &DL, SDValue Chain, SDValue Dst, SDValue Src,
+    SDValue Size, RTLIB::Libcall LC) const {
+  const AArch64Subtarget &STI =
+      DAG.getMachineFunction().getSubtarget<AArch64Subtarget>();
+  const AArch64TargetLowering *TLI = STI.getTargetLowering();
+  TargetLowering::ArgListTy Args;
+  TargetLowering::ArgListEntry Entry;
+  SDValue Symbol;
+  Entry.Ty = DAG.getDataLayout().getIntPtrType(*DAG.getContext());
+  Entry.Node = Dst;
+  Args.push_back(Entry);
+  EVT Ty = TLI->getPointerTy(DAG.getDataLayout());
+
+  switch (LC) {
+  case RTLIB::MEMCPY:
+    Symbol = DAG.getExternalSymbol("__arm_sc_memcpy", Ty);
+    Entry.Node = Src;
```

Comment:
I'm not a fan of defining `Entry.Node` once above, and then partially re-defining it here for a different purpose. For example, this assumes Entry.Ty is set correctly for the uses here and below.
As another example: it sets `Entry.IsSExt` to `false` when `LC == RTLIB::MEMSET`, which is also used for the `Size` argument. But when `LC != RTLIB::MEMSET`, `Entry.IsSExt` is `true`. That means the Size argument will either be sign-extended or not, depending on the value of `LC`. I'm not sure this is intentional.

Reply:
hmm, I think it should be not for size last parameter, but for charecter parameter and forbiding to sign-extends looks correct for me.

Reply:
> I think it should be not for the size last parameter(n)

From looking at the code though, it seems that Size will be zero-extended only for memset, and not for memmove/memcpy. The way you've written this code by partially re-defining `Entry` is prone to possible bugs and this looks like it may be one of them.

Reply:
it is not the Size argument that we do zero-extend/trunc operation but 
Charecter (int c) only for memset second parameter only.  the last argument(Size) should be intact for all three operations.

Reply:
I will rearrange the code.

Reply:
oh, I see the issue here. Thanks Sander for finding this one.

Reply:
Can you rewrite this as follows:

```
case ....: {
  TargetLowering::ArgListEntry Entry;
  ..
  Entry.Node = Src;
  Entry.Ty = PtrTy;
  Args.push_back(Entry);
  break;
}
```

That way, we don't accidentically use any previously set values in `Entry` that don't apply to this operand.